rfc-170-mobile-app-backend-stack.md

```diff
@@ -0,0 +1,80 @@
+
+As GOV.UK, we need to determine a tech stack for back end services which underpin the GOV.UK elements of the app.
+
+The following are key considerations we need to take into account:
+
+1. There is a single app, and it would be most straightforward to integrate the app with a single back end. This means we only need a single set of configuration rather than multiple settings for different features. Additionally, we can centralise concerns like logging, monitoring and alerting.
+1. In all likelihood, the various mobile back end services across both DI and GOV.UK will _eventually_ end up being supported and maintained by the same group of people. Therefore, parity of skillset is an advantage.
+1. There are broader conversations across GDS looking into our strategy around tech stacks, and we should fold in to those
+
+## Options
+
+We have considered two principal options for tech stacks:
+
+1. Adopt a serverless stack on AWS
+2. Adopt a Ruby-based containerised stack on AWS
+
+### Option 1 - Adopt a serverless stack on AWS
+
+![Architecture diagram showing a simplified approach to running serverless applications on AWS](rfc-170/serverless.png)
+
+Under this option, we would adopt a serverless stack which broadly mirrors the approach taken on DI. Connectivity would be provided by API Gateways (see note below about the proposed approach to this). Compute would be based on Lambda functions, running the Node.JS runtime.
+
+The Lambda functions would be able to call out to other AWS services as needed. While the subject of this RFC is predominantly the connectivity and compute (i.e. API Gateway and Lambda), the proposal would also be to adopt other cloud-native services which commonly form part of a serverless approach, namely using DynamoDB for storage.
+
+This option has the following advantages:
+* Essentially unlimited scalability without the need for active management
+* Cost-efficient, pay-for-what-you-use pricing (we can get some cost reports from DI if this would be helpful)
+* A step towards no-ops, where we are not responsible for or have the cost/burden of infra maintenance as this would be handled by AWS, so there are no [tricky manual steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) to upgrade to latest versions of underpinning software (only deprecated runtimes need to be considered)
+* Lambda functions are small and self-contained, thus easy to reason about. They are written in JavaScript, which is familiar to many developers and therefore the learning curve is not too steep
+* Aligns with the DI tech stack so that the entire mobile back end estate could be maintained by a single group of people
+* Benefit from existing DI infrastructure for NFRs like monitoring / logging / alerting
```

Comment:
These are advantages if the app sits close to DI in the org structure. They're disadvantages if it sits close to GOV.UK (e.g. if GOV.UK needed to support this infrastructure using its existing structures, it would be problematic to have to learn new monitoring / logging tools)